src/Types/Animations/Animation.tsx

```diff
@@ -8,66 +8,92 @@ import { WobblySwarm } from './WobblySwarm';
 
 export function pushNewFigure(selectedFigure, figs, p: p5) {
+    figs.push(newFigure(selectedFigure, p.mouseX, p.mouseY, p));
+}
+
+export function newFigure(selectedFigure, x: number, y:number, p: p5, color?:string) {
     let dimension = Math.random() * 50 + 20;
-    let color;
 
     switch(selectedFigure) {
         case SelectedShape.Circle:
-            color = Animation.circleColors[Math.floor(Math.random() * Animation.circleColors.length)];
-            let newCirc = new CircleFigure(p.mouseX, p.mouseY, dimension, color, p);
-            figs.push(newCirc);
-            break;
+            if (!color) color = Animation.circleColors[Math.floor(Math.random() * Animation.circleColors.length)];
+            return new CircleFigure(x, y, dimension, color, p);
             
         case SelectedShape.Rectangle:
-            color = Animation.rectColors[Math.floor(Math.random() * Animation.rectColors.length)];
-            let newSquare = new SquareFigure(p.mouseX, p.mouseY, dimension, color, p);
-            figs.push(newSquare);
-            break;
+            if (!color) color = Animation.rectColors[Math.floor(Math.random() * Animation.rectColors.length)];
+            return new SquareFigure(x, y, dimension, color, p);
 
         case SelectedShape.Triangle:
-            color = Animation.triangleColors[Math.floor(Math.random() * Animation.triangleColors.length)]
-            let newTriangle = new TriangleFigure(p.mouseX, p.mouseY, dimension, color, p);
-            figs.push(newTriangle);
-            break;
+            if (!color) color = Animation.triangleColors[Math.floor(Math.random() * Animation.triangleColors.length)]
+            return new TriangleFigure(x, y, dimension, color, p);
+        default:
     }
 }
-
+    
 export class Animation {
 
     static rectColors = [];
     static circleColors = [];
     static triangleColors = [];
 
+    static freeUndefinedFigures(sketchData: SketchData) {
+        for (let i = 0; i < sketchData.figs.length; i++) {
+            if (sketchData.figs[i] === undefined) {
+                sketchData.figs.splice(i, 1);
+            }
+        }
+    }
+
     static propsHandler(sketchData, p) {
         Animation.rectColors = generateColorSpectrum(sketchData.colorSettings.rectangle);
         Animation.circleColors = generateColorSpectrum(sketchData.colorSettings.circle);
         Animation.triangleColors = generateColorSpectrum(sketchData.colorSettings.triangle);
     }
 
+    static redraw(sketchData: SketchData, p: p5) {
+        switch(sketchData.selectedAnimation) {
+            case SelectedAnimation.FillScreenWithFigures:
+                    FillScreenWithFigures.redraw(sketchData, p);
+                break;
+            default:
+                p.frameRate(60);
+        }
+    }
+
     static draw(sketchData: SketchData, p) {
+
+        // This is still a problem sometimes ???? 
+        this.freeUndefinedFigures(sketchData); 
```

Comment:
Figures are sometimes stored in json as undefined objects. When we load them back in from the browser and try to use them, they start breaking other things. This might not be the best way to handle this case, so I'll look at it again.

Reply:
I should also look around and see why objects are being stored as undefined, but my guess is that the browser is getting refreshed right in the middle of the call to serialize the figures in the draw loop. -- If that is the case, the only fix would be do something like what I'm doing now.

Reply:
Nvm, the change for downward gravity fixed this.